File: leetcode/C++/path-sum-iii.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int sol;
    int curSum;
    unordered_map<int,int> um;
    void traverse(TreeNode *node, int &targetSum){
        if(node==nullptr) return;
        curSum+=node->val;
        if(curSum==targetSum) ++sol;
        sol+=um[curSum-targetSum];
        ++um[curSum];
        traverse(node->left, targetSum);
        traverse(node->right, targetSum);
        --um[curSum];
        curSum-=node->val;
    }
    int pathSum(TreeNode* root, int targetSum) {
        sol=curSum=0;
        traverse(root,targetSum);
        return sol;
    }
};

/*
o(n^2) 풀이는 떠올리기 쉬웠으나 o(n)은 떠올리기 쉽지 않았음.
*/

File: leetcode/Python/path-sum-iii.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> int:
        sol, um, curSum = [0], dict(), [0]
        def traverse(node):
            if node is None:
                return
            curSum[0] += node.val
            if curSum[0] == targetSum:
                sol[0]+=1
            sol[0] += um.get(curSum[0]-targetSum, 0)
            um[curSum[0]]=um.get(curSum[0], 0) + 1
            traverse(node.left)
            traverse(node.right)
            um[curSum[0]]-=1
            curSum[0] -= node.val
        traverse(root)
        return sol[0]